File: 300/306 - Additive Number.java
import java.math.BigInteger;

public class Solution {
    public boolean isAdditiveNumber(String num) {
        int i, j, len = num.length();
        
        for (i = 0; i < len - 2; ++i) {
            for (j = i + 1; j < len - 1; ++j) {
                if (dfs(num.substring(0, i + 1), num.substring(i + 1, j + 1), num.substring(j + 1)))
                    return true;
            }
        }
        
        return false;
    }
    
    private boolean dfs(String a, String b, String c) {
        String r;
        int i;
            
        if (Math.max(a.length(), b.length()) > c.length() || (! isValid(a)) || (! isValid(b)))
            return false;
        
        r = new BigInteger(a).add(new BigInteger(b)).toString();
        
        if (r.length() > c.length())
            return false;
        
        for (i = 0; i < r.length(); ++i) {
            if (r.charAt(i) != c.charAt(i))
                return false;
        }
        
        return r.length() == c.length() ? true : dfs(b, r, c.substring(r.length()));
    }
        
    private boolean isValid(String s) {
        return s.length() == 1 || s.charAt(0) != '0';
    }
}

File: csharp/306.additive-number.cs
/*
 * @lc app=leetcode id=306 lang=csharp
 *
 * [306] Additive Number
 */
public class Solution {

    private bool IsAdditiveNumberRecurse(string num, string n1, string n2) {
        if ((n1.Length > 1 && n1[0] == '0') || (n2.Length > 1 && n2[0] == '0')) {
            return false;
        }
        string sum = (long.Parse(n1) + long.Parse(n2)).ToString();
        if (num.Length <= sum.Length) {
            return num == sum;
        }
        if (num.Substring(0, sum.Length) != sum) {
            return false;
        }
        return IsAdditiveNumberRecurse(num.Substring(sum.Length), n2, sum);
    }

    public bool IsAdditiveNumber(string num) {
        int n = num.Length;
        for (int i = 1; i < n; ++i) {
            string n1 = num.Substring(0, i);
            for (int j = 1; j < n - i; ++j) {
                string n2 = num.Substring(i, j);
                if (IsAdditiveNumberRecurse(num.Substring(i + j), n1, n2)) {
                    return true;
                }
            }
        }
        return false;
    }
}

